File: articles/cognitive-services/Speech-Service/code/translation_samples.cs
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//

// <toplevel>
using System;
using System.Media;
using System.IO;
using System.Collections.Generic;
using System.Threading.Tasks;
using Microsoft.CognitiveServices.Speech;
using Microsoft.CognitiveServices.Speech.Translation;
// </toplevel>

namespace MicrosoftSpeechSDKSamples
{
    public class TranslationSamples
    {
        // Translation from microphone.
        // <TranslationWithMicrophoneAsync>
        public static async Task TranslationWithMicrophoneAsync()
        {
            // Creates an instance of a speech factory with specified
            // subscription key and service region. Replace with your own subscription key
            // and service region (e.g., "westus").
            var factory = SpeechFactory.FromSubscription("YourSubscriptionKey", "YourServiceRegion");

            // Sets source and target languages
            string fromLanguage = "en-US";
            List<string> toLanguages = new List<string>() { "de" };

            // Sets voice name of synthesis output.
            const string GermanVoice = "de-DE-Hedda";

            // Creates a translation recognizer using microphone as audio input, and requires voice output.
            using (var recognizer = factory.CreateTranslationRecognizer(fromLanguage, toLanguages, GermanVoice))
            {
                // Subscribes to events.
                recognizer.IntermediateResultReceived += (s, e) => {
                    Console.WriteLine($"\nPartial result: recognized in {fromLanguage}: {e.Result.RecognizedText}.");
                    if (e.Result.TranslationStatus == TranslationStatus.Success)
                    {
                        foreach (var element in e.Result.Translations)
                        {
                            Console.WriteLine($"    Translated into {element.Key}: {element.Value}");
                        }
                    }
                    else
                    {
                        Console.WriteLine($"    Translation failed. Status: {e.Result.TranslationStatus.ToString()}, FailureReason: {e.Result.FailureReason}");
                    }
                };

                recognizer.FinalResultReceived += (s, e) => {
                    if (e.Result.RecognitionStatus != RecognitionStatus.Recognized)
                    {
                        Console.WriteLine($"\nFinal result: Status: {e.Result.RecognitionStatus.ToString()}, FailureReason: {e.Result.RecognitionFailureReason}.");
                        return;
                    }
                    else
                    {
                        Console.WriteLine($"\nFinal result: Status: {e.Result.RecognitionStatus.ToString()}, recognized text in {fromLanguage}: {e.Result.RecognizedText}.");
                        if (e.Result.TranslationStatus == TranslationStatus.Success)
                        {
                            foreach (var element in e.Result.Translations)
                            {
                                Console.WriteLine($"    Translated into {element.Key}: {element.Value}");
                            }
                        }
                        else
                        {
                            Console.WriteLine($"    Translation failed. Status: {e.Result.TranslationStatus.ToString()}, FailureReason: {e.Result.FailureReason}");
                        }
                    }
                };

                recognizer.SynthesisResultReceived += (s, e) =>
                {
                    if (e.Result.Status == SynthesisStatus.Success)
                    {
                        Console.WriteLine($"Synthesis result received. Size of audio data: {e.Result.Audio.Length}");
                        using (var m = new MemoryStream(e.Result.Audio))
                        {
                            SoundPlayer simpleSound = new SoundPlayer(m);
                            simpleSound.PlaySync();
                        }
                    }
                    else if (e.Result.Status == SynthesisStatus.SynthesisEnd)
                    {
                        Console.WriteLine($"Synthesis result: end of synthesis result.");
                    }
                    else
                    {
                        Console.WriteLine($"Synthesis error. Status: {e.Result.Status.ToString()}, Failure reason: {e.Result.FailureReason}");
                    }
                };

                recognizer.RecognitionErrorRaised += (s, e) => {
                    Console.WriteLine($"\nAn error occurred. Status: {e.Status.ToString()}");
                };

                recognizer.OnSessionEvent += (s, e) => {
                    Console.WriteLine($"\nSession event. Event: {e.EventType.ToString()}.");
                };

                // Starts continuous recognition. Uses StopContinuousRecognitionAsync() to stop recognition.
                Console.WriteLine("Say something...");
                await recognizer.StartContinuousRecognitionAsync().ConfigureAwait(false);

                do
                {
                    Console.WriteLine("Press Enter to stop");
                } while (Console.ReadKey().Key != ConsoleKey.Enter);

                await recognizer.StopContinuousRecognitionAsync().ConfigureAwait(false);
            }
        }
        // </TranslationWithMicrophoneAsync>

        // Translation using file input.
        // <TranslationWithFileAsync>
        public static async Task TranslationWithFileAsync()
        {
            // Creates an instance of a speech factory with specified
            // subscription key and service region. Replace with your own subscription key
            // and service region (e.g., "westus").
            var factory = SpeechFactory.FromSubscription("YourSubscriptionKey", "YourServiceRegion");

            // Sets source and target languages
            string fromLanguage = "en-US";
            List<string> toLanguages = new List<string>() { "de", "fr" };

            var translationEndTaskCompletionSource = new TaskCompletionSource<int>();

            // Creates a translation recognizer using file as audio input.
            // Replace with your own audio file name.
            using (var recognizer = factory.CreateTranslationRecognizerWithFileInput(@"YourAudioFile.wav", fromLanguage, toLanguages))
            {
                // Subscribes to events.
                recognizer.IntermediateResultReceived += (s, e) => {
                    Console.WriteLine($"\nPartial result: recognized in {fromLanguage}: {e.Result.RecognizedText}.");
                    if (e.Result.TranslationStatus == TranslationStatus.Success)
                    {
                        foreach (var element in e.Result.Translations)
                        {
                            Console.WriteLine($"    Translated into {element.Key}: {element.Value}");
                        }
                    }
                    else
                    {
                        Console.WriteLine($"    Translation failed. Status: {e.Result.TranslationStatus.ToString()}, FailureReason: {e.Result.FailureReason}");
                    }
                };

                recognizer.FinalResultReceived += (s, e) => {
                    if (e.Result.RecognitionStatus != RecognitionStatus.Recognized)
                    {
                        Console.WriteLine($"\nFinal result: Status: {e.Result.RecognitionStatus.ToString()}, FailureReason: {e.Result.RecognitionFailureReason}.");
                        return;
                    }
                    else
                    {
                        Console.WriteLine($"\nFinal result: Status: {e.Result.RecognitionStatus.ToString()}, recognized text in {fromLanguage}: {e.Result.RecognizedText}.");
                        if (e.Result.TranslationStatus == TranslationStatus.Success)
                        {
                            foreach (var element in e.Result.Translations)
                            {
                                Console.WriteLine($"    Translated into {element.Key}: {element.Value}");
                            }
                        }
                        else
                        {
                            Console.WriteLine($"    Translation failed. Status: {e.Result.TranslationStatus.ToString()}, FailureReason: {e.Result.FailureReason}");
                        }
                    }
                };

                recognizer.RecognitionErrorRaised += (s, e) => {
                    Console.WriteLine($"\nAn error occurred. Status: {e.Status.ToString()}");
                };

                recognizer.OnSpeechDetectedEvent += (s, e) => {
                    Console.WriteLine($"\nSpeech detected event. Event: {e.EventType.ToString()}.");
                    // Stops translation when speech end is detected.
                    if (e.EventType == RecognitionEventType.SpeechEndDetectedEvent)
                    {
                        Console.WriteLine($"\nStop translation.");
                        translationEndTaskCompletionSource.TrySetResult(0);
                    }
                };

                // Starts continuous recognition. Uses StopContinuousRecognitionAsync() to stop recognition.
                Console.WriteLine("Start translation...");
                await recognizer.StartContinuousRecognitionAsync().ConfigureAwait(false);

                // Waits for completion.
                await translationEndTaskCompletionSource.Task.ConfigureAwait(false);

                // Stops translation.
                await recognizer.StopContinuousRecognitionAsync().ConfigureAwait(false);
            }
        } 
        // </TranslationWithFileAsync>
    }
}

File: articles/cognitive-services/Speech-Service/code/translation_samples.cpp
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//

#include "stdafx.h"

// <toplevel>
#include <string>
#include <vector>
#include <speechapi_cxx.h>

using namespace std;
using namespace Microsoft::CognitiveServices::Speech;
using namespace Microsoft::CognitiveServices::Speech::Translation;
// </toplevel>


// Translation with microphone input.
// <TranslationWithMicrophone>
void TranslationWithMicrophone()
{
    // Creates an instance of a speech factory with specified
    // subscription key and service region. Replace with your own subscription key
    // and service region (e.g., "westus").
    auto factory = SpeechFactory::FromSubscription(L"YourSubscriptionKey", L"YourServiceRegion");

    // Sets source and target languages
    auto fromLanguage = L"en-US";
    vector<wstring> toLanguages { L"de", L"fr" };

    // Creates a translation recognizer using microphone as audio input.
    auto recognizer = factory->CreateTranslationRecognizer(fromLanguage, toLanguages);
    wcout << L"Say something...\n";

    // Starts translation. It returns when the first utterance has been recognized.
    auto result = recognizer->RecognizeAsync().get();

    // Checks result.
    if (result->Reason != Reason::Recognized)
    {
        wcout << L"There was an error in speech recognition, reason " << int(result->Reason) << L"-" << result->ErrorDetails << '\n';
    }
    else if (result->TranslationStatus != TranslationStatusCode::Success)
    {
        wcout << L"There was an error in translation, status: " << int(result->TranslationStatus) << '\n';
    }
    else
    {
        wcout << "We recognized in " << fromLanguage << ": " << result->Text << '\n';
        for (const auto& it : result->Translations)
        {
            wcout << L"    Translated into " << it.first.c_str() << ": " << it.second.c_str() << '\n';
        }
    }
}
// </TranslationWithMicrophone>

// Translation with file input.
// <TranslationWithFile>
void TranslationWithFile()
{
    // Creates an instance of a speech factory with specified
    // subscription key and service region. Replace with your own subscription key
    // and service region (e.g., "westus").
    auto factory = SpeechFactory::FromSubscription(L"YourSubscriptionKey", L"YourServiceRegion");

    // Sets source and target languages
    auto fromLanguage = L"en-US";
    vector<wstring> toLanguages{ L"de", L"fr" };

    // Creates a translation recognizer using file as audio input.
    // Replaces with your own audio file name.
    auto recognizer = factory->CreateTranslationRecognizerWithFileInput(L"YourAudioFile.wav", fromLanguage, toLanguages);

    // Starts translation. It returns when the first utterance has been recognized.
    auto result = recognizer->RecognizeAsync().get();

    // Checks result.
    if (result->Reason != Reason::Recognized)
    {
        wcout << L"There was an error in speech recognition, reason " << int(result->Reason) << L"-" << result->ErrorDetails << '\n';
    }
    else if (result->TranslationStatus != TranslationStatusCode::Success)
    {
        wcout << L"There was an error in translation, status: " << int(result->TranslationStatus) << '\n';
    }
    else
    {
        wcout << "We recognized in " << fromLanguage << ": " << result->Text << '\n';
        for (const auto& it : result->Translations)
        {
            wcout << L"    Translated into " << it.first.c_str() << ": " << it.second.c_str() << '\n';
        }
    }
}
// </TranslationWithFile>

// Defines event handlers for different events.
// <TranslationContinuousRecognitionUsingEvents>
static void OnPartialResult(const TranslationTextResultEventArgs& e)
{
    wcout << L"IntermediateResult: Recognized text:" << e.Result.Text << '\n';
    for (const auto& it : e.Result.Translations)
    {
        wcout << L"    Translated into " << it.first.c_str() << ": " << it.second.c_str() << '\n';
    }
}

static void OnFinalResult(const TranslationTextResultEventArgs& e)
{
    wcout << L"FinalResult: status:" << (int)e.Result.TranslationStatus << L". Recognized Text: " << e.Result.Text << '\n';
    for (const auto& it : e.Result.Translations)
    {
        wcout << L"    Translated into " << it.first.c_str() << ": " << it.second.c_str() << '\n';
    }
}

static void OnSynthesisResult(const TranslationSynthesisResultEventArgs& e)
{
    if (e.Result.SynthesisStatus == SynthesisStatusCode::Success)
    {
        wcout << L"Translation synthesis result: size of audio data: " << e.Result.Audio.size();
    }
    else if (e.Result.SynthesisStatus == SynthesisStatusCode::Error)
    {
        wcout << L"Translation synthesis error: " << e.Result.FailureReason;
    }
}

static void OnNoMatch(const TranslationTextResultEventArgs& e)
{
    wcout << L"NoMatch:" << (int)e.Result.Reason << '\n';
}

static void OnCanceled(const TranslationTextResultEventArgs& e)
{
    wcout << L"Canceled:" << (int)e.Result.Reason << L"- " << e.Result.Text << '\n';
}

// Continuous translation.
void TranslationContinuousRecognitionUsingEvents()
{
    // Creates an instance of a speech factory with specified
    // subscription key and service region. Replace with your own subscription key
    // and service region (e.g., "westus").
    auto factory = SpeechFactory::FromSubscription(L"YourSubscriptionKey", L"YourServiceRegion");

    // Sets source and target languages
    auto fromLanguage = L"en-US";
    vector<wstring> toLanguages{ L"de", L"fr" };

    // Creates a translation recognizer using microphone as audio input.
    auto recognizer = factory->CreateTranslationRecognizer(fromLanguage, toLanguages);

    // Subscribes to events.
    recognizer->IntermediateResult.Connect(&OnPartialResult);
    recognizer->FinalResult.Connect(&OnFinalResult);
    recognizer->Canceled.Connect(&OnCanceled);
    recognizer->NoMatch.Connect(&OnNoMatch);
    recognizer->TranslationSynthesisResultEvent.Connect(&OnSynthesisResult);

    wcout << L"Say something...\n";

    // Starts continuos recognition. Uses StopContinuousRecognitionAsync() to stop recognition.
    recognizer->StartContinuousRecognitionAsync().wait();

    wcout << L"Press any key to stop\n";
    string s;
    getline(cin, s);

    // Stops recognition.
    recognizer->StopContinuousRecognitionAsync().wait();

    // Unsubscribes to events.
    recognizer->IntermediateResult.Disconnect(&OnPartialResult);
    recognizer->FinalResult.Disconnect(&OnFinalResult);
    recognizer->Canceled.Disconnect(&OnCanceled);
    recognizer->NoMatch.Disconnect(&OnNoMatch);
    recognizer->TranslationSynthesisResultEvent.Disconnect(&OnSynthesisResult);
}
// </TranslationContinuousRecognitionUsingEvents>
